Depth-first search for dependency resolution

lindsayad opened the pull request:
I will see about combining this with the code we added in #20778 so we don't have code/concept duplication.

Closes #19856

andrsd committed:
Using depth-first search for dependency resolution

Refs #0

lindsayad committed:
Add const version of MaterialPropertyStorage::statefulProps

lindsayad committed:
Only resize our material properties in PorousFlowMaterial

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1053678/) on fc26c6b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21108/site)

This comment will be updated on new commits.

lindsayad committed:
Allow an object to depend on itself

This is our pre-existing behavior and it's necessary for use of
discrete materials as objects in which properties are iteratively
computed.

Additionally, we use a multimap because multiple objects can supply
an item, e.g. different materials on different blocks can supply
the same material property

lindsayad committed:
Replace DependencyResolver class

cpgr commented:
Thanks @lindsayad, it looks like our hacky material resizing still works with this change. I hope it didn't cause you too many headaches.

lindsayad committed:
Create DependencyResolver 2.0 based on David's graph

lindsayad commented:
No it wasn't too bad! I always enjoy getting some time in other folks' modules and getting some peripheral sense of what they're doing. It reminds me of how awesome of a community we have and how thankful I am for external contributors like yourself and @WilkAndy

lindsayad committed:
Add Syntax::addDontInsertDependency API

This allows addDependency to continue to work as a dependency
*insertion* tool, keeping the same conceptual behavior as before

lindsayad committed:
Use same code for doing both dfs and cycle determination

lindsayad committed:
Iterate by insertion order

This is necessary for being in sync in multi-process runs. This
has a cost associated with it, e.g. when we are iterating over
the insertion order to begin our depth-first search we have to
perform O(log n) look-ups in the adjacency map, but hopefully
this will not show up as too much of a penalty

lindsayad committed:
More nuanced handling of insertDependency

Now if we detect that the key is already in the graph, then we
early return instead of failing assertions. We check to make sure
that there is a depends-on relationship, e.g. an edge from value
to key, before early returning

lindsayad committed:
Use DFS in THM Simulation.C

lindsayad committed:
Try a deque for insertion ordering and push_front

lindsayad committed:
Throw cyclic exception directly from dfs

lindsayad committed:
Revamp insertDependency and removeDependenciesOf

- insertDependency is back to being just a straight addEdge
- removeDependenciesOf keeps reciprociy and removes corresponding
  items in the adjacencies data member

lindsayad committed:
Remove DependencyResolver::operator()

lindsayad committed:
Revert "Try a deque for insertion ordering and push_front"

This reverts commit 6dfbd6c1f3ac38000e26af89ae361e8827ff2096.

lindsayad committed:
Add newly necessary edges

We are no longer using insertDependency to make the dependents of
value the dependents of key, which necessitates adding some new
edges manually

lindsayad committed:
Update DependencyResolver unit testing

We update testing to only test whether we have had a correct
topological sort. We do not make any assumptions about root
traversal order

lindsayad committed:
Update THM to new dependency resolver

lindsayad committed:
Update tensor mechanics to new dependency resolver

lindsayad committed:
Catch MetaPhysicL exceptions in ThreadedElementLoopBase

lindsayad committed:
Update FP and NS to new dependency resolver

lindsayad committed:
Update porous flow to new dependency resolver

lindsayad committed:
heat conduction dependency resolver update

lindsayad committed:
geochemistry dependency resolver update

lindsayad committed:
stochastic tools dependency resolver update

lindsayad committed:
contact dependency resolver update

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1062923/) on 2fed89f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>97b8b2</th>
      <th colspan=4>#21108 2fed89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.58%</td>
      <td>82.58%</td>
      <td>-0.00%</td>
      <td>95.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75531</td>
      <td>75487</td>
      <td>-44</td>
      <td>144</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15932</td>
      <td>15926</td>
      <td>-6</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>97b8b2</th>
      <th colspan=4>#21108 2fed89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>89.01%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2876</td>
      <td>2876</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>355</td>
      <td>355</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage/contact)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>97b8b2</th>
      <th colspan=4>#21108 2fed89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.28%</td>
      <td>96.28%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10532</td>
      <td>10545</td>
      <td>+13</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>407</td>
      <td>407</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage/porous_flow)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>97b8b2</th>
      <th colspan=4>#21108 2fed89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.02%</td>
      <td>90.03%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4538</td>
      <td>4544</td>
      <td>+6</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>503</td>
      <td>503</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage/stochastic_tools)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>97b8b2</th>
      <th colspan=4>#21108 2fed89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.16%</td>
      <td>91.16%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12012</td>
      <td>12015</td>
      <td>+3</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1165</td>
      <td>1165</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21108/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21108/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented:
@lindsayad No new syntax tasks. I used `IterationAdaptiveDT`, so a different number of time steps would result if the number of nonlinear iterations was ever different.

joshuahansel commented:
Looks like 108 instead of 99 time steps for the open Brayton test. Do you get 108 steps with the framework changes alone, or only when you apply THM changes?

lindsayad commented:
The dependency resolver changes *do* result in different topological sorts. The old resolver, for instance, would put all "independent" items at the front of the sorted vector. However, with the new resolver, you may get an independent item way later in the sorted vector, after all the items that you traversed up from the first root you examined.

I noticed that for the shaft connected turbine/compressor tests that I re-golded, your energy conservation got much better. I'm still pursuing the question of whether the old dependency resolver didn't actually sort user objects correctly

lindsayad commented:
For the shaft tests, I explicitly added the indirect dependencies of the `FunctionValuePostprocessor`. It just so happened that even without adding those, the new dependency resolver gave the same sorting. However, with the previous dependency resolver, I'm confident that the `FunctionValuePostprocessor` was executing before some of the other postprocessors and so then had old information, resulting in the bad energy conservation.

I didn't see something obvious like that in the Brayton inputs unfortunately ☹️

joshuahansel commented:
> So with the new resolver, you are experiencing worse convergence?

I haven't looked at the new transient. It's possible that less iterations on one time step could cause a growth and lead to a failure, leading to a cutback and overall more time steps.

joshuahansel commented:
I can't find anything that seems relevant. I think we'll need to compare the time step histories to see what's going on.

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1073607/) on d48b558 : invalidated by @lindsayad

lindsayad committed:
Make sure that we have same task ordering as prior in FP

lindsayad committed:
Move methods outside class definition and add some more doxygen

lindsayad committed:
Remove unused methods

lindsayad commented:
This should be ready for review

joshuahansel commented on modules/thermal_hydraulics/src/components/FlowChannelBase.C:
```diff
@@ -299,6 +299,9 @@ FlowChannelBase::buildMesh()
   }
+
+  // Update the mesh
+  _mesh.update();
```
Just curious - what's this for?

GiudGiud commented on modules/porous_flow/src/base/PorousFlowApp.C:
```diff
@@ -66,13 +66,15 @@ associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
   // Task dependency and syntax for action to automatically add PorousFlowJoiner materials
   registerTask("add_joiners", /*is_required=*/false);
   addTaskDependency("add_joiners", "add_material");
+  addTaskDependency("add_master_action_material", "add_joiners");
```
how did you determine this whole list of extra dependencies?
Was this just to make the test pass or do you have a list that you impose on each module?

lindsayad replied:
See previous comment. In general the new task would get inserted immediately behind the prereq task in the task list given in `Moose.C`. If you look in `Moose.C`, `add_master_action_material` is the task that occurs right after `add_material`

GiudGiud commented on test/tests/misc/check_error/tests:
```diff
@@ -737,7 +737,7 @@
     type = 'RunException'
     input = 'scalar_old_integrity_check.i'
-    expect_err = 'Vector tag \'TIME\' does not exist'
+    expect_err = 'Calling \'coupledScalarValueOld\' on variable \S+ when using a \"Steady\" executioner is not allowed. This value is available only in transient simulations.'
```
love the improved error messages!

GiudGiud reviewed (commented):
should be good after this

lindsayad commented on modules/navier_stokes/src/base/NavierStokesApp.C:
```diff
@@ -53,34 +53,42 @@ associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
   // add variables action
   registerTask("add_navier_stokes_variables", /*is_required=*/false);
   addTaskDependency("add_navier_stokes_variables", "add_variable");
+  addTaskDependency("add_mortar_variable", "add_navier_stokes_variables");
```
The old dependency resolver used to insert a new task immediately behind its prereq, and in front of other things that depend on the prereq ... but not always. In fact what it did was not very transparent at all. With the new dependency resolver you explicitly add every dependency edge; it's fully transparent and obvious what is happening. Now does `add_mortar_variable` truly depend on `add_navier_stokes_variables`? I would sure hope not! But where does the dependency appear? It is hard for me to say except through Monte Carlo. It should definitely happen before `add_user_object`; but should it happen before `add_fluid_properties` for example? I don't actually know unless I try and then maybe I get an error in the FP module. Maybe I don't, but then an error crops up in SAM. What I do know is that if I glance at the old sorting of tasks in the debugger, `add_navier_stokes_variables` occurred before `add_mortar_variable` and so I know this is a safe location for it in the graph.

GiudGiud replied:
Interesting. This seems like there should be convention when there is indetermination on the order. Like "alphabetical sorting"
adding dependencies makes the graph larger but since your algorithm is faster we dont notive it

GiudGiud replied:
but adding dependencies also makes the graph more consistent across apps, which is good

lindsayad committed:
Update framework/include/interfaces/DependencyResolverInterface.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>
